bfd/elfxx-riscv.c

```diff
@@ -1294,21 +1324,154 @@ riscv_parse_sv_or_non_std_ext (riscv_parse_subset_t *rps,
 
       *q = '\0';
 
+      /* Check that the name is valid.
+	 For 'x', anything goes but it cannot simply be 'x'.
+	 For 'z', it must be known from a list and also cannot simply be 'z'.
+	 For 's', it must be known from a list and also *can* simply be 's'.
+	 For 'sx', see 'x'.  */
+
+      /* Check that the extension name is well-formed.  */
+      if (!config->ext_valid_p (subset))
+	{
+	  rps->error_handler
+	    ("-march=%s: Invalid or unknown %s ISA extension: '%s'",
+	     march, config->prefix, subset);
+	  free (subset);
+	  return NULL;
+	}
+
+      /* Check that the last item is not the same as this.  */
+      last_name = rps->subset_list->tail->name;
+
+      if (!strcasecmp (last_name, subset))
+	{
+	  rps->error_handler ("-march=%s: Duplicate %s ISA extension: \'%s\'",
+			      march, config->prefix, subset);
+	  free (subset);
+	  return NULL;
+	}
+
+      /* Check that we are in alphabetical order within the subset.  */
+      if (!strncasecmp (last_name, config->prefix, 1)
+	  && strcasecmp (last_name, subset) > 0)
+	{
+	  rps->error_handler ("-march=%s: %s ISA extension not in alphabetical order: "
+			      "\'%s\' must come before \'%s\'.",
+			      march, config->prefix, subset, last_name);
+	  free (subset);
+	  return NULL;
+	}
+
       riscv_add_subset (rps->subset_list, subset, major_version, minor_version);
       free (subset);
       p += end_of_version - subset;
 
       if (*p != '\0' && *p != '_')
 	{
 	  rps->error_handler ("-march=%s: %s must seperate with _",
-			      march, ext_type_str);
+			      march, config->prefix);
 	  return NULL;
 	}
     }
 
   return p;
 }
 
+const char * const riscv_std_z_ext_strtab[] =
+  {
+   "zbb", "zbc", "zbe", "zbf", "zbm", "zbp", "zbr", "zbs", "zbt", NULL
+  };
+
+const char * const riscv_std_s_ext_strtab[] =
+  {
+   NULL
+  };
+
+/* Find the index corresponding to the z-extension
+   name `ext'. If not found, return -1.  */
+
+int
+riscv_std_z_ext_index (const char *ext)
+{
+  int i;
+
+  for (i = 0; riscv_std_z_ext_strtab[i]; ++i)
+    {
+      if (!strcasecmp (ext, riscv_std_z_ext_strtab[i]))
+	return i;
+    }
+
+  return -1;
+}
+
+/* Predicator function for x-prefixed extensions.
+   Anything goes, except the literal 'x'.  */
+
+static bfd_boolean
+riscv_ext_x_valid_p (const char *arg)
+{
+  if (!strcasecmp (arg, "x"))
+    return FALSE;
+
+  return TRUE;
+}
+
+/* Predicator functions for z-prefixed extensions.
+   Only known z-extensions are permitted.  */
+
+static bfd_boolean
+riscv_ext_z_valid_p (const char *arg)
```

Comment:
However I think riscv_ext_z_valid_p and riscv_ext_s_valid_p need to using some common utils function to check, otherwise we'll have 3 almost copy/paste code after supporting h* extension.

like
```
static bfd_boolean
riscv_ext_z_valid_p (const char *arg)
{
   return riscv_multi_letter_ext_valid_p (arg, riscv_z_ext_strtab);
}
static bfd_boolean
riscv_ext_s_valid_p (const char *arg)
{
   return riscv_multi_letter_ext_valid_p (arg, riscv_s_ext_strtab);
}
```